lib/llm/src/key_value_store/etcd.rs

```diff
@@ -0,0 +1,245 @@
+            .map_err(|e| StorageError::EtcdError(e.to_string()))?;
+        let out: HashMap<String, bytes::Bytes> = resp
+            .into_iter()
+            .map(|kv| {
+                let (k, v) = kv.into_key_value();
+                (String::from_utf8_lossy(&k).to_string(), v.into())
+            })
+            .collect();
+
+        Ok(out)
+    }
+}
+
+impl EtcdBucket {
+    async fn create(&self, key: &str, value: &str) -> Result<StorageOutcome, StorageError> {
+        let k = make_key(&self.endpoint, &self.bucket_name, key);
+        tracing::trace!("etcd create: {k}");
+
+        // Does it already exists? For 'create' it shouldn't.
+        let kvs = self
+            .client
+            .kv_get(k.clone(), None)
+            .await
+            .map_err(|e| StorageError::EtcdError(e.to_string()))?;
+        if !kvs.is_empty() {
+            let version = kvs.first().unwrap().version();
+            return Ok(StorageOutcome::Exists(version as u64));
+        }
+
+        // Write it
+        let mut put_resp = self
+            .client
+            .kv_put(k, value, Some(PutOptions::new().with_prev_key()))
+            .await
+            .map_err(|e| StorageError::EtcdError(e.to_string()))?;
+        // Check if we overwrite something
+        if put_resp.take_prev_key().is_some() {
+            // Key created between our get and put
+            return Err(StorageError::Retry);
+        }
+
+        // version of a new key is always 1
+        Ok(StorageOutcome::Created(1))
+    }
+
+    async fn update(
+        &self,
+        key: &str,
+        value: &str,
+        revision: u64,
+    ) -> Result<StorageOutcome, StorageError> {
+        let version = revision;
+        let k = make_key(&self.endpoint, &self.bucket_name, key);
+        tracing::trace!("etcd update: {k}");
+
+        let kvs = self
+            .client
+            .kv_get(k.clone(), None)
+            .await
+            .map_err(|e| StorageError::EtcdError(e.to_string()))?;
+        if kvs.is_empty() {
+            return Err(StorageError::MissingKey(key.to_string()));
+        }
+        let current_version = kvs.first().unwrap().version() as u64;
+        if current_version != version + 1 {
+            tracing::warn!(
+                current_version,
+                attempted_next_version = version,
+                key,
+                "update: Wrong revision"
+            );
+            // NATS does a resync_update, overwriting the key anyway and getting the new revision.
+            // So we do too in etcd.
+        }
+
+        let mut put_resp = self
+            .client
+            .kv_put(k, value, Some(PutOptions::new().with_prev_key()))
+            .await
+            .map_err(|e| StorageError::EtcdError(e.to_string()))?;
+        Ok(match put_resp.take_prev_key() {
+            // Should this be an error?
+            // The key was deleted between our get and put. We re-created it.
+            None => StorageOutcome::Created(1),
```

Comment:
1. Is this magic number 1 the default version for a newly created key? Does a const make sense instead?

2. An inline comment somewhere linking to the etcd docs on this might be nice: https://etcd.io/docs/v3.5/learning/data_model/

> The key space maintains multiple revisions. When the store is created, the **initial revision is 1.** 
> ...
> A key’s life spans a generation, from creation to deletion. Each key may have one or multiple generations. Creating a key increments the version of that key, starting at 1 if the key does not exist at the current revision. Deleting a key generates a key tombstone, concluding the key’s current generation by resetting its version to 0

3. Do we do this tombstone version 0 thing on delete anywhere?

Reply:
Added the comment, link and clarification. 

1 isn't a magic number, it's the first revision of something. I think delete has a flag not revision 0, but not 100%, etcd library handles that.